x-pack/legacy/plugins/maps/public/layers/vector_layer.js

```diff
@@ -484,12 +490,84 @@ export class VectorLayer extends AbstractLayer {
     }
   }
 
+  async _syncSourceFormatters(syncContext) {
+    if (this._style.constructor.type !== LAYER_STYLE_TYPE.VECTOR) {
+      return;
+    }
+
+    return this._syncFormatters({
+      source: this._source,
+      dataRequestId: SOURCE_FORMATTERS_ID_ORIGIN,
+      dynamicStyleProps: this._style.getDynamicPropertiesArray().filter(dynamicStyleProp => {
+        return dynamicStyleProp.getFieldOrigin() === FIELD_ORIGIN.SOURCE;
+      }),
+      ...syncContext,
+    });
+  }
+
+  async _syncJoinFormatters(syncContext, join) {
+    const joinSource = join.getRightJoinSource();
+    return this._syncFormatters({
+      source: joinSource,
+      dataRequestId: join.getSourceFormattersDataRequestId(),
+      dynamicStyleProps: this._style.getDynamicPropertiesArray().filter(dynamicStyleProp => {
+        const matchingField = joinSource.getMetricFieldForName(
+          dynamicStyleProp.getField().getName()
+        );
+        return dynamicStyleProp.getFieldOrigin() === FIELD_ORIGIN.JOIN && !!matchingField;
+      }),
+      ...syncContext,
+    });
+  }
+
+  async _syncFormatters({
+    source,
+    dataRequestId,
+    dynamicStyleProps,
+    startLoading,
+    stopLoading,
+    onLoadError,
+  }) {
+    if (dynamicStyleProps.length === 0) {
+      return;
+    }
+
+    const fieldNames = dynamicStyleProps.map(dynamicStyleProp => {
+      return dynamicStyleProp.getField().getName();
+    });
+    const nextMeta = {
+      fieldNames: _.uniq(fieldNames).sort(),
+    };
+    const prevDataRequest = this._findDataRequestForSource(dataRequestId);
+    const canSkipUpdate = canSkipFormattersUpdate({ prevDataRequest, nextMeta });
+    if (canSkipUpdate) {
+      return;
+    }
+
+    const requestToken = Symbol(`layer-${this.getId()}-${dataRequestId}`);
+    try {
+      startLoading(dataRequestId, requestToken, nextMeta);
+
+      const formatters = {};
+      const promises = dynamicStyleProps.map(async dynamicStyleProp => {
```

Comment:
`fields.map` directly iso the indirection over style-properties.

Reply:
still can not use `map` here. Results of getFieldFormatter is a field formatter but you have lost the field name so you don't know how to map it. Still need `fields.forEach`